File: frontend/src/store/modules/BulletinBoard.js
// initial state
const state = {
  electionID: null,
  voters: null,
  candidates: null,
  titles: null,
  partialPublicVotingCredentials: null,
  numberOfSelections: null,
  numberOfCandidates: null,
  publicKey: null,
  countingCircles: null,
  eligibilityMatrix: null,
  numberOfParallelElections: null,
  publicKeyShares: [],
  ballots: [],
  confirmations: [],
  encryptions: [],
  decryptions: [],
  shuffleProofs: [],
  decryptionProofs: []
}

// mutations
const mutations = {
  SOCKET_SYNCBULLETINBOARD: (state, data) => {
    const bb = JSON.parse(data)
    state.voters = bb.voters
    state.candidates = bb.candidates
    state.titles = bb.titles
    state.partialPublicVotingCredentials = bb.partialPublicVotingCredentials
    state.numberOfSelections = bb.numberOfSelections
    state.numberOfCandidates = bb.numberOfCandidates
    state.publicKeyShares = bb.publicKeyShares
    state.publicKey = bb.publicKey
    state.countingCircles = bb.countingCircles
    state.eligibilityMatrix = bb.eligibilityMatrix
    state.numberOfParallelElections = bb.numberOfParallelElections
    state.ballots = bb.ballots
    state.confirmations = bb.confirmations
    state.encryptions = bb.encryptions
    state.decryptions = bb.decryptions
    state.shuffleProofs = bb.shuffleProofs
    state.decryptionProofs = bb.decryptionProofs
    state.verificationResult = bb.verificationResult
  }
}

const getters = {
  getBallotsOfBulletinBoard: (state, getters) => {
    // returns the ballotlist of the bulletin board
    return state.ballots
  },
  getBallotById: (state, getters) => (ballotId) => {
    for (let ballot of getters.getBallotsOfBulletinBoard) {
      if (ballot.id === ballotId) {
        return {ballot: ballot}
      }
    }
  },
  getConfirmationById: (state, getters) => (confirmationId) => {
    for (let confirmation of getters.getConfirmationsOfBulletinBoard) {
      if (confirmation.id === confirmationId) {
        return {confirmation: confirmation}
      }
    }
  },
  getConfirmationsOfBulletinBoard: (state, getters) => {
    // returns the confirmation list of the bulletin board
    return state.confirmations
  },
  getDecryptionsForAuthority: (state, getters) => (authorityId) => {
    // returns the list of decryptions of a given election authority
    if (state.decryptions.length < authorityId + 1) { return null }
    return state.decryptions[authorityId]
  },
  hasDecryptionTask: (state, getters) => (authorityId) => {
    // returns true or false depending on whether the passed election authority has a pending decryption task
    if (getters.status !== 5) return 0
    if (authorityId === 0) { return getters.getDecryptionsForAuthority(authorityId) === null } else {
      return getters.getDecryptionsForAuthority(authorityId) === null && getters.getDecryptionsForAuthority(authorityId - 1) !== null
    }
  },
  haveAllAuthoritiesDecrypted: (state, getters) => {
    // returns true or false depending on whether all authorities have finished decrypting
    return getters.getDecryptionsForAuthority(getters.numberOfElectionAuthorities - 1) !== null
  }
}

export default {
  state,
  mutations,
  getters
}

File: backend/app/actors/BulletinBoard.py
from chvote.ElectionAuthority.GenElectorateData import GenElectorateData
from app.actors.Actor import Actor
from chvote.Utils.Utils import AssertList, AssertInt, AssertMpz, AssertClass
from chvote.Types import VerificationResult

class BulletinBoard(Actor):

    def __init__(self, collection, electionID):
        Actor.__init__(self, collection, electionID)

    @property
    def securityLevel(self):
        return self.state.securityLevel

    @securityLevel.setter
    def securityLevel(self, value):
        AssertInt(value)
        self.state.securityLevel = value

    @property
    def voters(self):
        return self.state.voters

    @voters.setter
    def voters(self, value):
        AssertList(value)
        self.state.voters = value


    @property
    def titles(self):
        return self.state.titles

    @titles.setter
    def titles(self, value):
        AssertList(value)
        self.state.titles = value


    @property
    def countingCircles(self):
        return self.state.countingCircles

    @countingCircles.setter
    def countingCircles(self, value):
        AssertList(value)
        self.state.countingCircles = value

    @property
    def candidates(self):
        return self.state.candidates

    @candidates.setter
    def candidates(self, value):
        AssertList(value)
        self.state.candidates = value

    @property
    def numberOfCandidates(self):
        return self.state.numberOfCandidates

    @numberOfCandidates.setter
    def numberOfCandidates(self, value):
        AssertList(value)
        self.state.numberOfCandidates = value

    @property
    def numberOfSelections(self):
        return self.state.numberOfSelections

    @numberOfSelections.setter
    def numberOfSelections(self, value):
        AssertList(value)
        self.state.numberOfSelections = value

    @property
    def numberOfParallelElections(self):
        return self.state.numberOfParallelElections

    @numberOfParallelElections.setter
    def numberOfParallelElections(self, value):
        AssertInt(value)
        self.state.numberOfParallelElections = value

    @property
    def eligibilityMatrix(self):
        return self.state.eligibilityMatrix

    def generateEligibilityMatrix(self):
        self.state.eligibilityMatrix = [[True for el in range(self.numberOfParallelElections)] for i in range(len(self.voters))]

    @property
    def partialPublicVotingCredentials(self):
        return self.state.partialPublicVotingCredentials

    @partialPublicVotingCredentials.setter
    def partialPublicVotingCredentials(self, value):
        AssertList(value)
        self.state.partialPublicVotingCredentials = value

    @property
    def ballots(self):
        return self.state.ballots

    @ballots.setter
    def ballots(self, value):
        AssertList(value)
        self.state.ballots = value

    @property
    def confirmations(self):
        return self.state.confirmations

    @confirmations.setter
    def confirmations(self, value):
        AssertList(value)
        self.state.confirmations = value

    @property
    def publicKeyShares(self):
        return self.state.publicKeyShares

    @publicKeyShares.setter
    def publicKeyShares(self, value):
        AssertList(value)
        self.state.publicKeyShares = value

    @property
    def publicKey(self):
        return self.state.publicKey

    @publicKey.setter
    def publicKey(self, value):
        AssertMpz(value)
        self.state.publicKey = value

    @property
    def encryptions(self):
        return self.state.encryptions

    @encryptions.setter
    def encryptions(self, value):
        AssertList(value)
        self.state.encryptions = value

    @property
    def shuffleProofs(self):
        return self.state.shuffleProofs

    @shuffleProofs.setter
    def shuffleProofs(self, value):
        AssertList(value)
        self.state.shuffleProofs = value

    @property
    def decryptions(self):
        return self.state.decryptions

    @decryptions.setter
    def decryptions(self, value):
        AssertList(value)
        self.state.decryptions = value


    @property
    def decryptionProofs(self):
        return self.state.decryptionProofs

    @decryptionProofs.setter
    def decryptionProofs(self, value):
        AssertList(value)
        self.state.decryptionProofs = value

    @property
    def verificationResult(self):
        return self.state.verificationResult

    @verificationResult.setter
    def verificationResult(self, value):
        AssertClass(value, VerificationResult)
        self.state.verificationResult = value


    def getBallotById(self, ballotId):
        for ballot in self.ballots:
            if ballot.id == ballotId:
                return ballot
        return None


    def getConfirmationById(self, confirmationId):
        for confirmation in self.confirmations:
            if confirmation.id == confirmationId:
                return confirmation
        return None